plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c

```diff
@@ -0,0 +1,274 @@
+fu_plugin_thunderbolt_power_get_path (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autoptr(GList) devices = NULL;
+	const gchar *basepath;
+	const gchar *driver;
+
+	/* in case driver went away */
+	if (data->force_path != NULL)
+		g_free (data->force_path);
+
+	devices = g_udev_client_query_by_subsystem (data->udev, "wmi");
+	for (GList* l = devices; l != NULL; l = l->next) {
+		g_autofree gchar *built_path = NULL;
+		GUdevDevice *device = l->data;
+
+		/* only supports intel-wmi-thunderbolt for now */
+		driver = g_udev_device_get_driver (device);
+		if (g_strcmp0 (driver, "intel-wmi-thunderbolt") != 0)
+			continue;
+
+		/* check for the attribute to be loaded */
+		basepath = g_udev_device_get_sysfs_path (device);
+		if (basepath == NULL)
+			continue;
+		built_path = g_build_path ("/", basepath,
+					   "force_power", NULL);
+		if (g_file_test (built_path, G_FILE_TEST_IS_REGULAR)) {
+			data->force_path = g_steal_pointer (&built_path);
+			g_debug ("Detected force power support at %s",
+				 data->force_path);
+			break;
+		}
+	}
+	g_list_foreach (devices, (GFunc) g_object_unref, NULL);
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_supported (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	return (data->force_path != NULL);
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_set (FuPlugin *plugin, gboolean enable,
+				 GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	gint fd;
+	gint ret;
+
+	if (!fu_plugin_thunderbolt_power_supported (plugin)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "unable to set power to %d (missing kernel support)",
+			     enable);
+		return FALSE;
+	}
+	g_debug ("Setting force power to %d", enable);
+	fd = g_open (data->force_path, O_WRONLY);
+	if (fd == -1) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "failed to open %s", data->force_path);
+		return FALSE;
+	}
+	ret = write (fd, enable ? "1" : "0", 1);
+	if (ret < 1) {
+		g_set_error (error, G_IO_ERROR,
+			     g_io_error_from_errno (errno),
+			     "could not write to force_power': %s",
+			     g_strerror (errno));
+		g_close (fd, NULL);
+		return FALSE;
+	}
+	return g_close (fd, error);
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_reset_cb (gpointer data)
+{
+	if (!fu_plugin_thunderbolt_power_set (FU_PLUGIN (data), FALSE, NULL))
+		g_warning ("failed to reset thunderbolt power");
+	return FALSE;
+}
+
+static gboolean
+udev_uevent_cb (GUdevClient *udev,
+		const gchar *action,
+		GUdevDevice *device,
+		gpointer     user_data)
+{
+	FuPlugin *plugin = FU_PLUGIN(user_data);
+
+	if (action == NULL)
+		return TRUE;
+
+	g_debug ("uevent for %s: %s", g_udev_device_get_sysfs_path (device), action);
+
+	/* intel-wmi-thunderbolt has been loaded/unloaded */
+	if (g_str_equal (action, "change")) {
+		fu_plugin_thunderbolt_power_get_path (plugin);
+		fu_plugin_set_enabled (plugin, TRUE);
```

Comment:
`TRUE` or only if `get_path` succeeded finding it? I mean, it needs to be disabled for unload, isn't it?

Reply:
It should get disabled from the recoldplug,  but you're right may as well not schedule a re-coldplug if it's not going to be supported.